core/primitives/src/sandbox.rs

```diff
@@ -0,0 +1,73 @@
+#[cfg(feature = "sandbox")]
+pub mod state_patch {
+    use crate::state_record::StateRecord;
+
+    /// Changes to the state to be applied via sandbox-only state patching
+    /// functionality.
+    ///
+    /// As we only expose this functionality for sandbox, we make sure that the
+    /// object is only constructable if sandbox feature is enabled.
+    #[derive(Default)]
+    pub struct SandboxStatePatch {
+        records: Vec<StateRecord>,
+    }
+
+    impl SandboxStatePatch {
+        pub fn new(records: Vec<StateRecord>) -> SandboxStatePatch {
+            SandboxStatePatch { records }
+        }
+
+        pub fn is_empty(&self) -> bool {
+            self.records.is_empty()
+        }
+
+        pub fn clear(&mut self) {
+            self.records.clear();
+        }
+
+        pub fn take(&mut self) -> SandboxStatePatch {
+            Self { records: core::mem::take(&mut self.records) }
+        }
+
+        pub fn merge(&mut self, other: SandboxStatePatch) {
+            self.records.extend(other.records);
+        }
+    }
+
+    impl IntoIterator for SandboxStatePatch {
+        type Item = StateRecord;
+        type IntoIter = std::vec::IntoIter<StateRecord>;
+
+        fn into_iter(self) -> Self::IntoIter {
+            self.records.into_iter()
+        }
+    }
+}
+
+#[cfg(not(feature = "sandbox"))]
+pub mod state_patch {
+    use crate::state_record::StateRecord;
+
+    #[derive(Default)]
+    pub struct SandboxStatePatch;
+
+    impl SandboxStatePatch {
+        pub const fn is_empty(&self) -> bool {
+            true
+        }
+        pub const fn clear(&self) {}
+        pub fn take(&mut self) -> Self {
+            Self
+        }
+        pub const fn merge(&self, _other: SandboxStatePatch) {}
```

Comment:
```suggestion
        pub fn is_empty(&self) -> bool {
            true
        }
        pub fn clear(&self) {}
        pub fn take(&mut self) -> Self {
            Self
        }
        pub fn merge(&self, _other: SandboxStatePatch) {}
```

let's remove `const` to make API the same, to avoid the problem when the code works only without sandbox. I would generally prefer to follow `don't cfg API` rule here, to make sure by construction that callers won't get compile-time errors. So,

```rust
mod state_patch  {
    pub struct SandboxStatePatch {
        #[cfg(feature = "sandbox")]
        records: Vec<StateRecord>,
    }
}
```

mild preference for this. 

For `IntoIterator`, I think its worth it to break the rule and to use different types for `IntoIterator`, as that would directly affect how the compiler sees the code. 


If the `const` here is intended just as a hint to the compiler that this is a no-op, I don't think it'd work that way (but @nagisa is the expert). You'd rather want to mark that as `#[inline]`, but even that is not really needed as we use full lto for pod build

Reply:
While it would make it easier to maintain compatible APIs, it would also force peppering the code with many more `#[cfg(…)]` and `if `cfg! { }` lines for, IMO, a decrease in readability.

Reply:
True!